File: src/main/java/org/kite9/diagram/common/algorithms/fg/AbsoluteArc.java
package org.kite9.diagram.common.algorithms.fg;

/**
 * A type of arc where positive or negative flow is permitted, within a boundary, 
 * but any flow in either direction has a constant cost.
 * @author Rob Moffat
 *
 */
public class AbsoluteArc extends AbstractArc {

	protected int cost;
		
	protected int capacity;

	public AbsoluteArc(int cost, int capacity, Node from, Node to, String label) {
		super();
		this.cost = cost;
		this.capacity = capacity;
		this.from = from;
		this.to = to;
		this.label = label;
		from.getArcs().add(this);
		to.getArcs().add(this);
	}

	public int getFlowCost() {
		return Math.abs(flow) * cost;
	}

	public int getIncrementalCost(int flow) {
		int origCost = Math.abs(this.flow)*cost;
		int newCost = Math.abs(this.flow+flow) *cost;
		return newCost - origCost;
	}

	public boolean hasCapacity(boolean reversed) {
		if (isBlocked()) {
			return false;
		}
		if (!reversed) {
			return flow < capacity;
		} else {
			return flow > -capacity;
		}
	}
	
}

File: src/main/kotlin/org/kite9/diagram/common/algorithms/fg/AbsoluteArc.kt
package org.kite9.diagram.common.algorithms.fg

/**
 * A type of arc where positive or negative flow is permitted, within a boundary,
 * but any flow in either direction has a constant cost.
 * @author Rob Moffat
 */
open class AbsoluteArc(var cost: Int, protected var capacity: Int, from: Node, to: Node, label: String) :
    AbstractArc(from, to, label, 0) {

    open fun getFlowCost(): Int {
        return Math.abs(flow) * cost
    }

    override fun getIncrementalCost(flow: Int): Int {
        val origCost = Math.abs(this.flow) * cost
        val newCost = Math.abs(this.flow + flow) * cost
        return newCost - origCost
    }

    override fun hasCapacity(reversed: Boolean): Boolean {
        return if (!reversed) {
            flow < capacity
        } else {
            flow > -capacity
        }
    }
}